### PR TITLE
Gets additional data from ArchivesSpace, returns dictionary

### PR DESCRIPTION
**Pull request by bonniegee**
Gets data from AS as defined in #33. Uses the _resolved parameter.

I'm not sure if how I've handled instances is the best way of approaching it. Right now, I first check to see if there is a digital object or a microform instance. If there is a microform instance, I return that container only, and if there are not microform instances or digital objects, I return only the container for the first instance. In addition to posing issues if we have archival objects with multiple instances of the same type, it assume that any instance type that is not "microform" should be handled in the same way as "mixed materials."

For dates, I only return the first date; since multiple dates are an outlier (especially below the collection level) and because of how this data will be used, I figured this was the best approach.

Additionally, I left placeholder functions which did not work with the _resolved parameter approach.

Does not tackle restrictions yet--wanted to push the rest up first, since that may take some time

### CODE REVIEW

**Comment by p-galligan** [comment]
So, I think you may have done more than you needed to right now for instances/containers. It looks like the logic here is set up to get the preferred delivery format based on the instances. In conversations with @helrond , I believe we may want to expand the `as_data` fixture to include a list of all instance container type and indicator concatenated by comma as a string, and then a separate field as the preferred delivery format. It looks like you've done that second part mostly already.

**Comment by bonniegee** [comment]
Thanks @p-galligan! If we want to get all container types and indicators, do we also want to get all locations, or do we only want the location for the preferred delivery format?

**Comment by p-galligan** [comment]
Only the location for the preferred delivery format, if it has one.

**Comment by helrond** [comment]
Wow this is actually pretty well fleshed out. I agree with what @p-galligan said: we need a preferred delivery container and then a field with all the containers. So maybe we add `preferred_container` and `preferred_location` (just to make clear they are the same thing) and change `container` to `containers`. However, I think working more on that is really part of. #20 and #21 aka on @p-galligan's plate.

I have to admit I haven't dug into this in great depth, but I'm wondering if `check_for_instance_type` shouldn't just return the instance(s) which match the instance type provided? I think that would be slightly more efficient. 

I think you could easily account for multiple dates with another for loop/list comprehension. I agree it's an edge case, but it's an easy one to cover.

A couple of other thoughts:
- I see you're calling obj.json() a lot - you may want to just set that to a variable to clean things up a bit.
- For the helper functions, you may want to think about passing in just the data you need, rather than the entire json blob. For. example, pass in the resolved linked agents from the resource to `get_collection_creator` rather than doing all the key selection in the first line there. I think that might open up the path to some further refactoring in the helpers, and they could potentially become more abstract and reusable.

**Comment by bonniegee** [comment]
@p-galligan It might take me a bit to get the restrictions data; would it be helpful to get this into a mergeable state sooner (and file the restrictions data work as a separate issue) so that work on #20 and #21 aren't blocked?

**Comment by p-galligan** [comment]
@bonniegee Yeah, that’d be perfect.

**Comment by bonniegee** [comment]
I pushed up some more changes, including making some of the helpers more targeted. I did not modify the helper for getting instances, since I wanted to leave all the container stuff alone. I modified the fixtures to include data with the `_resolved` parameter, but I did not add additional tests since I got stuck on how to use fixture data in the middle of a method.

**Review by p-galligan — APPROVED** [review]
I'll probably make some changes to the instances sections when I dig into my issues, but otherwise looks good!

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]:
```suggestion
    """Gets a human-readable location string for a top container
    
    Args:
        top_container_info (dict): json for a top container (with resolved container locations)
        
     Returns:
         string: all locations associated with the top container, separated by a comma.
     """
```

Comment (helrond) [reply process_request/helpers.py]: Can all docstrings be formatted as above - i.e. provide a one-line summary of the function, and break out args and returns?

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: this *could* be done in one line via list comprehension:
```suggestion
    locations = [c.get("_resolved").get("title") for c in top_container_info.get("container_locations")]
```

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: See comment about docstrings above.

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Should this method be called "get_container_field" or something like that? I find the current name to be a little misleading...

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: See comment about docstrings above.

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Errrr, unless I am missing something the logic inside both of these cases is identical...

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Again, an opportunity for list comprehension (which you can always do if you're just performing one action inside a for loop:
```suggestion
        dates = [get_expression(d) for d in archival_object.get("dates")]
```

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Docstring formatting

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: don't you want to be passing in `item_collection` here?

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: I could be wrong about this, but I believe that an empty list is falsy, so you should be able to just do 
```suggestion
            if item_json.get("ancestors"):
```

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: This is probably stuff that @p-galligan will pick up, but this logic should probably be moved out o this function into a separate method that's called something like "get_preferred".

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Docstring formatting (see fuller comment below)

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: if you pass in the resource json directly it will save you some trouble here...see my comment below in the `get_data` method.

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: this falls under @p-galligan's area of work but I think this function should just return the relevant instances, rather than an index which you then have to use in another method to get the instance.

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: The logic inside both of these statements is exactly the same?

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: This method does not appear to be called; can we delete it?

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: Again, this is stuff that @p-galligan will be working on, but i think this should all get put in another method called "get_preferred" or something like that.

**Review by helrond — CHANGES_REQUESTED** [review]
This is looking good! An overall request to improve docstring formatting (and consistency) and then a bunch of other relatively minor things.

File: process_request/helpers.py
Comment (bonniegee) [reply process_request/helpers.py]: I left this alone since I didn't know if it would be helpful for @p-galligan's work.

Comment (p-galligan) [reply process_request/helpers.py]: I think we can delete it. I can always add it back if it becomes necessary. However, if we're using the resolve approach, I don't think it will be.

File: process_request/routines.py
Comment (p-galligan) [reply process_request/routines.py]: Yeah. I'm going to make some changes to instance, container, and location logic here. I'm happy to just pull stuff out for now or wait until after this get merged.

**Comment by p-galligan** [comment]
@helrond @bonniegee I want to limit how much work Bonnie does to "clean up" sections that I'll be working on and changing anyway. What would you prefer for those three comments that pertain to "my" sections?

File: process_request/helpers.py
Comment (helrond) [reply process_request/helpers.py]: There's already a `get_locations` method that, based on the docstrings, should do the same thing as this, and works with resolved data rather than making additional fetches.

**Comment by helrond** [comment]
@p-galligan my comments relating to your suggestions were meant to be FYIs for you, not additional changes for @bonniegee.

File: process_request/routines.py
Comment (bonniegee) [reply process_request/routines.py]: A collection is an ancestor, which is why I was checking that there's more than 1 ancestor.

**Comment by bonniegee** [comment]
This still needs tests before it's merged, but I think this is in a state where you can branch off of it to work on #20 and #21 @p-galligan

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Is this method really necessary? All you're doing is calling `get()` on a dict...

File: process_request/helpers.py
Comment (helrond) [inline process_request/helpers.py]: Can we remove this code from this PR? It's a little hard to evaluate what changes are actually being made otherwise.

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: Is there a reason we're not using `None`?

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: Same as above = can this get removed from this PR so we can more easily see the changes that are being proposed?

File: process_request/routines.py
Comment (helrond) [inline process_request/routines.py]: I think it might be a bit cleaner to NOT set an empty dict to a variable, but instead to just do:

```
return {
  "creator": get_collection_creator(item_collection),
  "restrictions": "TK,
...
}
```
It's a bit of a personal preference thing but I feel like it might clarify things a bit in the code.

File: process_request/tests.py
Comment (helrond) [inline process_request/tests.py]: Instead of asserting it's not False, I'd suggest asserting the return value is a dict
```suggestion
        self.assertTrue(isinstance(get_as_data, dict))
```

**Review by helrond — CHANGES_REQUESTED** [review]
Main request is just to yank out code you're not actually using so we can more clearly see the changes being proposed here.

**Review by helrond — APPROVED** [review]
This is looking very nice!

**Review by p-galligan — APPROVED** [review]

**Review by HaSistrunk — APPROVED** [review]

**Pull request merged**
